### PR TITLE
Ft naming and docs

### PR DESCRIPTION
**Pull request by maxspahn**
In this branch, we introduce type hinting and docstrings.

We use numpy docstrings as they tend to be most widely used..
Besides, we add the API documentation based on the docstrings to the sphinx documentation.

There should be no logic changes here. Only naming can be adapted and documentation added.

### CODE REVIEW

**Comment by maxspahn** [comment]
Addresses #35 
Still WIP

**Comment by maxspahn** [comment]
@alxschwrz @GijsGroote 
Let's stick to this docstring format for future code edits.
Quick reference in https://numpydoc.readthedocs.io/en/latest/format.html

I am not a big fan of documentation with docstrings, as I want to be the code self-explaining as much as possible.
Therefore, you don't need to always specify the return value if it is clear from the context. The same goes for parameters to functions/methods. Generally, keep it minimal.

I will where to put those commenting/style guidelines in the Github repository.

File: urdfenvs/urdfCommon/urdfEnv.py [hunk 1]
```diff
@@ -92,70 +169,132 @@ def step(self, action):
...
+
+        Parameters
+        ----------
 
-    def addSensor(self, sensor):
+        limits: Positions of walls, [[x_low, y_low], [x_high, y_high]]
+        """
+        colwallId = p.createCollisionShape(
+            p.GEOM_BOX, halfExtents=[0.05, 10.0, 0.5])
+        p.createMultiBody(
```
Comment (alxschwrz) [inline urdfenvs/urdfCommon/urdfEnv.py:239]: here a reason to use this fold out format here instead of the previous oneline function call? makes the code quite long.

**Review by alxschwrz — COMMENTED** [review]
@maxspahn I left some questions here, which were unclear to me. Other than those I reviewed the code and looks good to me. I also tested some example files. I needed to downgrade to pybullet version 3.2.0 in order to install all dependencies, which is not part of this PR though.

File: urdfenvs/urdfCommon/differentialDriveRobot.py [hunk 2]
```diff
@@ -99,22 +129,29 @@ def apply_torque_action(self, torques):
                 force=torques[i],
             )
 
-    def apply_acc_action(self, accs, dt):
-        self._vels_int += dt * accs
-        self._vels_int[0] = np.clip(
-            self._vels_int[0],
+    def apply_acceleration_action(self, accs: np.ndarray, dt: float) -> None:
```
[the remark this answers is not in the document]
Comment (maxspahn) [reply urdfenvs/urdfCommon/differentialDriveRobot.py:132]: exactly. It is just for editor easier to give you suggestions.

File: urdfenvs/urdfCommon/urdfEnv.py [hunk 3]
```diff
@@ -4,31 +4,97 @@
...
+from urdfenvs.urdfCommon.genericRobot import GenericRobot
 
 
 class WrongObservationError(Exception):
-    def __init__(self, msg, observation, observationSpace):
+    """Exception when observation lays outside the defined observation space.
+
+    This Exception is initiated when an the observation is not within the
+    defined observation space. The purpose of this exception is to give
+    the user better information about which specif part of the observation
```
[the remark this answers is not in the document]
Comment (maxspahn) [reply urdfenvs/urdfCommon/urdfEnv.py:18]: good catch!

**Comment by maxspahn** [comment]
As @alxschwrz is (rightfully!!) very keen on making the leading underscores correctly. I will change everything according to
https://peps.python.org/pep-0008/#method-names-and-instance-variables in an updated version.

TL;DR: non-public methods and class variables with leading underscore.

File: urdfenvs/urdfCommon/differentialDriveRobot.py [hunk 4]
```diff
@@ -158,24 +218,22 @@ def updateState(self):
...
         vf = np.array(
             [
-                0.5 * (v_right + v_left) * self._r,
-                (v_right - v_left) * self._r / self._l,
+                0.5 * (v_right + v_left) * self._wheelRadius,
+                (v_right - v_left) * self._wheelRadius / self._wheelDistance,
             ]
         )
-        Jnh = np.array(
+        jacobi_nonholonomic = np.array(
```
[the remark this answers is not in the document]
Comment (maxspahn) [reply urdfenvs/urdfCommon/differentialDriveRobot.py:233]: According to the PEP8 guideline, there is almost no case in which you use capitalized variables.

File: urdfenvs/urdfCommon/urdfEnv.py (hunk 1 above)
Comment (maxspahn) [reply urdfenvs/urdfCommon/urdfEnv.py:239]: The line would be 85 characters long without the folding.
That is too long according to PEP8 guidelines.

File: urdfenvs/mobileReacher/resources/mobileRobot.py [hunk 5]
```diff
@@ -25,17 +25,17 @@ def setJointIndices(self):
...
-    def setAccLimits(self):
+    def setAccelerationLimits(self):
         accLimit = np.array([1.0, 1.0, 1.0, 15.0, 7.5, 10.0, 12.5, 15.0, 20.0, 20.0, 1.0, 1.0])
         self._limitAcc_j[0, :] = -accLimit[0:self._n]
         self._limitAcc_j[1, :] = accLimit[0:self._n]
 
-    def apply_acc_action(self, accs, dt):
-        self._vels_int += dt * accs
-        self.apply_vel_action(self._vels_int)
+    def apply_acceleration_action(self, accs, dt):
```
[the remark this answers is not in the document]
Comment (GijsGroote) [reply urdfenvs/mobileReacher/resources/mobileRobot.py:36]: I disagree, but do want to take naming conventions to a higher level, a consistent naming style. 
for example: camelCase for python files, PascaleCase for classes and snake_case for functions.

**Review by GijsGroote — COMMENTED** [review]
I understand that at some point the documentation becomes too large to fit in the README.md. Now the readme contains information about the project, the documentation is in the documentation, Great job! Furthermore, developers could forget which style we use, this should be automated when pushing to GitHub for at least naming conventions.

File: .github/workflows/pylint.yml [hunk 6]
```diff
@@ -7,7 +7,7 @@ jobs:
     runs-on: ubuntu-latest
     strategy:
       matrix:
-        python-version: ["3.8", "3.9", "3.10"]
```
Comment (GijsGroote) [inline .github/workflows/pylint.yml:10]: Could be a reason for the pylint spitting out an error.

Comment (maxspahn) [reply .github/workflows/pylint.yml:10]: What exactly do you mean? the python 3.10?

File: urdfenvs/mobileReacher/resources/mobileRobot.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (maxspahn) [reply urdfenvs/mobileReacher/resources/mobileRobot.py:36]: Makes sense. We should agree on one style and stick to that everywhere.
So `apply_acceleration_action` does not fit in any style book, so that should be changed.
My  suggestion is:
Let's all have a read at the pep guide: https://realpython.com/python-pep8/#tips-and-tricks-to-help-ensure-your-code-follows-pep-8
Then we write a summary for this project in the repository description.
And then, I'll change everything in this PR accordingly.

**Comment by maxspahn** [comment]
I have found a very nice description of the Google style guide (it should be alright, it's google:D) 
https://google.github.io/styleguide/pyguide.html
They also offer a pylintrc configuration file that can be used with your texteditor. Then you will directly get warnings on the style while editing the files.
I will shortly add this to the repository.

**Comment by maxspahn** [comment]
I have changed everything according to the pep8 guidelines.
The full requirements are now stored in https://github.com/maxspahn/gym_envs_urdf/blob/ft-naming-and-docs/.pylintrc.
This is adapted from https://google.github.io/styleguide/pyguide.html.
Changes compared to the google style guide
- Ignore missing module docstrings
- Ignore missing class docstrings
- Default indent is 4 instead of 2 (2 is the pep8 guideline)

File: .github/workflows/pylint.yml (hunk 6 above)
Comment (GijsGroote) [reply .github/workflows/pylint.yml:10]: There was a pylint error, comment above points out that this _could be_ the source of the error. Because python 3.8, 3.9 and 3.10 are replaced by python 3.8
Since the pylint error is not present any more, this thread can be resolved

**Review by GijsGroote — APPROVED** [review]
Thank you for all the beautiful comments everywhere. I look forward to having to renaming because pylint tells me to. 👍🏼

**Comment by GijsGroote** [comment]
Ah, oke. With now 30 comments I felt that it became quite long. I will be more cautiously with merging in the future

**Comment by maxspahn** [comment]
I agree, but in this PR, almost every file is changed. This should be carefully reviewed.
